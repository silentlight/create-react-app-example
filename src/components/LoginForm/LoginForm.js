import React, { PureComponent } from 'react';
import { Field, reduxForm } from 'redux-form/immutable';
import PropTypes from 'prop-types';

import { required, email } from '../../utils/validations';

import { TextField } from 'components';

/**
 * Simple login form
 */
class LoginForm extends PureComponent {
  render() {
    const { onFormSubmit, isAuthenticating, handleSubmit, pristine, reset, submitting } = this.props;

    return (
      <form onSubmit={handleSubmit(onFormSubmit)}>

        <div>
          <Field
            name="email"
            component={TextField}
            type="email"
            label="Email"
            validate={[required, email]}
          />
        </div>

        <div>
          <Field
            name="password"
            component={TextField}
            type="password"
            label="Password"
            validate={required}
          />
        </div>

        <div>
          <button type="submit" disabled={pristine || submitting}>
            Submit
          </button>
          <button type="button" disabled={pristine || submitting} onClick={reset}>
            Clear Values
          </button>
        </div>

        {isAuthenticating &&
          <p>Authenticating user. Please wait...</p>
        }

      </form>
    );
  }
}

LoginForm.propTypes = {
  /** Callback to be run when form submits */
  onFormSubmit: PropTypes.func.isRequired,
  /** Indicator if form is authenticating */
  isAuthenticating: PropTypes.bool,
};

export default reduxForm({
  form: 'LoginForm',
})(LoginForm);
