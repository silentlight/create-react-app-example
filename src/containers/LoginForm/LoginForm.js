import React, { PureComponent } from 'react';
import { Field, reduxForm } from 'redux-form/immutable';
import PropTypes from 'prop-types';

import { required, email } from 'utils/validations';

import { TextField } from 'components';

/**
 * Simple login form
 */
export class LoginForm extends PureComponent {

  render() {
    const { submitting, pristine, reset, handleSubmit } = this.props;
    const { authenticating } = this.props;

    return (
      <form onSubmit={handleSubmit}>

        <div>
          <Field
            name="email"
            component={TextField}
            validate={[required, email]}
            type="email"
            label="Email"
          />
        </div>

        <div>
          <Field
            name="password"
            component={TextField}
            validate={required}
            type="password"
            label="Password"
          />
        </div>

        <div>
          <button type="submit" disabled={pristine || submitting}>
            Submit
          </button>
          <button type="button" disabled={pristine || submitting} onClick={reset}>
            Clear Values
          </button>
        </div>

        {authenticating &&
        <p>Authenticating user. Please wait...</p>
        }

      </form>
    );
  }
}

LoginForm.propTypes = {
  /** Callback to be run when form submits */
  onSubmit: PropTypes.func.isRequired,
  /** Indicator if form is authenticating */
  authenticating: PropTypes.bool,
};

export default reduxForm({
  form: 'LoginForm',
})(LoginForm);
