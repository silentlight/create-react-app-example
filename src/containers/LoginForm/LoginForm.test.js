import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import configureStore from 'config/initializers/store';

import LoginForm from './LoginForm';

const store = configureStore({});

describe('LoginForm', () => {

  const defaultProps = {
    pristine: true,
    submitting: false,
    handleSubmit: jest.fn(),
    reset: jest.fn(),

    authenticating: false,
    onSubmit: jest.fn(),
  };

  let component, props;

  const createComponent = (overrides = {}) => {
    props = Object.assign({}, defaultProps, overrides);

    return mount(
      <Provider store={store}><LoginForm {...props} /></Provider>
    );
  };

  it('renders and matches the snapshot', () => {
    const tree = renderer.create(<Provider store={store}><LoginForm {...defaultProps} /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('email field', () => {

    beforeEach(() => {
      component = createComponent();
    });

    it('is required', () => {
      component.find('[name="email"]').simulate('blur');
      expect(component.find('.warning').length).toEqual(1);
    });

    it('has to be valid email address', () => {
      component.find('input[name="email"]').simulate('change', { target: { value: 'test' } });
      component.find('[name="email"]').simulate('blur');
      expect(component.find('.warning').length).toEqual(1);

      component.find('input[name="email"]').simulate('change', { target: { value: 'test@test.com' } });
      expect(component.find('.warning').length).toEqual(0);
    });

  });

  describe('password field', () => {

    beforeEach(() => {
      component = createComponent();
    });

    it('is required', () => {
      component.find('[name="password"]').simulate('blur');
      expect(component.find('.warning').length).toEqual(1);
    })

  });

  describe('form submission', () => {

    beforeEach(() => {
      component = createComponent();
    });

    it("calls handleSubmit", () => {
      component.find('input[name="email"]').simulate('change', { target: { value: 'test@test.com' } });
      component.find('input[name="password"]').simulate('change', { target: { value: 'password' } });

      component.find('form').simulate('submit');

      expect(component.find('.warning').length).toEqual(0);
      expect(component.find('.error').length).toEqual(0);
      expect(props.handleSubmit.mock.calls.length).toEqual(1);
    })

  });

  describe('form reset', () => {

    beforeEach(() => {
      component = createComponent();
    });

    it("calls reset", () => {
      component.find('input[name="email"]').simulate('change', { target: { value: 'test@test.com' } });
      component.find('input[name="password"]').simulate('change', { target: { value: 'password' } });

      component.find('button[type="button"]').simulate('click');

      expect(component.find('input[name="email"]').prop('value')).toEqual('');
      expect(component.find('input[name="password"]').prop('value')).toEqual('');
    })

  });

  describe('authenticating prop', () => {

    beforeEach(() => {
      component = createComponent({ authenticating: true });
    });

    it("displays authentication text", () => {
      expect(component.html()).toContain('Authenticating user. Please wait...');
    })

  });

});