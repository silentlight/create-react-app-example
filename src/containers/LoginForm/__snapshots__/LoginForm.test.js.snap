// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders and matches the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <div>
      <label>
        Email
      </label>
      <div>
        <input
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
    </div>
  </div>
  <div>
    <div>
      <label>
        Password
      </label>
      <div>
        <input
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
  </div>
  <div>
    <button
      disabled={true}
      type="submit"
    >
      Submit
    </button>
    <button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Clear Values
    </button>
  </div>
</form>
`;
